### ISSUE TITLE
Pivot to unloaded root sections in initramfs

### ISSUE THREAD

**Issue by Winterhuman** [systemd/systemd#25428]
### Component

bootctl, systemd, systemd-boot, sd-stub

### Is your feature request related to a problem? Please describe

Root-embedded UKIs are an interesting use-case (https://github.com/uapi-group/specifications/discussions/32), however, one limitation with using an initramfs as the root FS is that it's stored in memory which can be quite significant for low-RAM systems.

### Describe the solution you'd like

Related: https://github.com/uapi-group/specifications/discussions/32#discussioncomment-4057510

To reduce the memory usage of large UKIs, PE sections can be marked as `noload` to prevent them from being put into RAM. Including a separate root FS in a UKI this way means it won't be loaded in memory, and from there systemd could then detect the unloaded root section, create a loopback device for it, and then pivot into the loopback device.

One issue with this approach is that the ESP would then have to remain mounted, ESP automount behaviour may need to be modified to not activate when the root is backed by a UKI in the ESP.

### Describe alternatives you've considered

(See edit history for the previous ideas)

### The systemd version you checked that didn't have the feature you are asking for

252

**Comment by medhefgo** [comment]
This sounds awfully complex (and brittle to me). Why not just use a small initramfs and a separate root image section and then let that initramfs look for the UKI and setup the root image inside to pivot into. By marking the root image section as noload it shouldn't be occupying memory after the image has been verified/hashed by the firmware.

This also means that no bootloader changes have to happen here (sd-stub should already be providing the needed info to userspace as to where the UKI was loaded from).

**Comment by Winterhuman** [comment]
I had no idea that you could mark sections as `noload`, that would work a lot better! Something similar was described in another discussion I opened (https://github.com/orgs/uapi-group/discussions/39), I guess this comes down to initramfs tooling then (e.g. `mkosi`)

**Comment by Winterhuman** [comment]
In that case, I guess support for this comes down to:

1. Searching for the unloaded root section in the UKI, and setting up a loopback (and `root=/dev/loop0` in the kernel cmdline to then pivot to it).
2. Generating the unloaded root section in the UKI, through tools like mkosi, dracut, etc.

**Comment by poettering** [comment]
does "noload" actually exist for PE binaries?

**Comment by medhefgo** [comment]
> does "noload" actually exist for PE binaries?

Yes. You can use `--set-section-flags section=noload` to set it. A quick look at binutils code shows it gets mapped to `IMAGE_SCN_MEM_DISCARDABLE ` PE section flag. Whether edk2/firmware makes use of that is a different matter though.

Mind you, embedding the root image in the initrd could get you OOM problems. The initrd is loaded into ram either way and then has to be copied into the final destination that the kernel gives us (it won't reuse whatever location it was loaded in). So your initrd/root image would have to fit twice in memory in that case.

**Comment by Winterhuman** [comment]
The closest evidence for `noload` being supported is this brief note in the edk2 docs: https://edk2-docs.gitbook.io/edk-ii-build-specification/2_design_discussion/26_creating_efi_images#text-2.6.4-generating-leaf-efi_section-files

Doesn't necessarily answer that UEFI firmware supports this though, right?

**Comment by medhefgo** [comment]
I don't see how a shared flag would help here.

The main problem is that if the root image is embedded in the efi binary it will be loaded into memory by firmware (most likely even if it is marked noload/discardable as firmware probably does not support this). So you already have a certain requirement on memory size.

The question is whether you can/want to afford this rootfs to be in memory twice because it is inside the initrd (and not its own section). This is because the kernel will not reuse the location of the initrd; it always expects to allocate the space itself and then tells us to copy it there.

I am simply saying that you really shouldn't embed it in the initrd and instead have the initrd find the rootfs section itself on the image on disk. Note that once the kernel exits EFI boot services the rootfs copy that was loaded by the firmware will simply be dropped on the floor. Marking the rootfs section as noload is really only a (theoretical) optimization for low memory machines.

**Comment by Winterhuman** [comment]
Yes, that's what I'm also saying; I changed the issue to mention using "unloaded root sections" when you first mentioned `noload` (potentially) being an option

**Comment by Winterhuman** [comment]
Ah right, I see where the confusion is, I still say the root section is "inside" the initramfs when I should be saying it's in the UKI. My mistake

**Comment by poettering** [comment]
so, what you could do: leave the initrd be. but add a new PE section ".rootfs" that is marked "noload" if that really exists. Then, use the initrd to find the kernel PE binary, find the ".rootfs" part of it, set up a loopback device from it, then transition into it. Done. That should give you the behaviour you want.

**Comment by Winterhuman** [comment]
@poettering Do you think it's reasonable to implement if systemd had native support for this? Both for auto-detecting the unloaded root section, and making it capable of creating loopback devices when a root section is found?
